### DISCUSSION TITLE
installing umqtt.simple package

### DISCUSSION THREAD

**Discussion by travelo94** [micropython/micropython#12082] [Using MicroPython]
Hello i have rpi pico that i want to connect with mqtt broker but i cant find a way to install 'socket' module , the connection with network will be via sim module 
`Traceback (most recent call last):
  File "<stdin>", line 3, in <module>
  File "umqtt/simple.py", line 1, in <module>
  File "/lib/usocket.py", line 2, in <module>
ImportError: no module named 'socket'`

**Comment by jimmo** [comment]
MicroPython does not currently have a socket-based implementation of a cellular radio driver. In general there is fairly limited support for cellular radios in MicroPython other than on ESP32 when using PPP (which will provides the socket module).

If you have a non-socket-based driver for your radio, it should be possible to write your own socket.py that adapts that into the standard socket interface. https://docs.micropython.org/en/latest/library/socket.html

**Reply by travelo94** [reply]
im using sim7000g , am i able establish constant communication with mqtt broker via AT commands?

**Reply by jimmo** [reply]
It should be possible to implement your own version of socket that does the AT commands. Then when umqtt does `import socket` it'll find your socket.py and use that.

Longer term, what we want to support is a way to defined custom network interfaces from Python. You can do this in C at the moment, see for example the [Nina driver](https://github.com/micropython/micropython/blob/master/extmod/network_ninaw10.c) which implements the [network driver protocol](https://github.com/micropython/micropython/blob/master/extmod/network_ninaw10.c#L828).

**Reply by travelo94** [reply]
i tried to send AT command to the sim7000module but it doesnt respondd is something in code that i write wrong ? / the tx of pico goes to rx of module and vice versa 

```py
from machine import Pin
import utime
import uos

uart0 = machine.UART(0,baudrate=9600)   #at-comand
def sendCMD_waitResp(cmd, uart=uart0, timeout=2000):
    print("CMD: " + cmd)
    uart.write(cmd)
    waitResp(uart, timeout)
    print()
    
def waitResp(uart=uart0, timeout=2000):
    prvMills = utime.ticks_ms()
    resp = b""
    while (utime.ticks_ms()-prvMills)<timeout:
        if uart.any():
            resp = b"".join([resp, uart.read()])
            
            if resp != '':
                print(resp.decode())   

   
sendCMD_waitResp("AT\r\n")
```

(Edited by @jimmo to add code formatting)

**Reply by jimmo** [reply]
Please see https://github.com/orgs/micropython/discussions/9111 (about how to format code for posting).

Also please use `time` and `os` rather than `utime` and `uos`. (This has been the advice for a long time but see https://github.com/orgs/micropython/discussions/12078 for recent updates about this)

**Reply by jimmo** [reply]
I notice that you aren't setting the tx and rx pins when creating the `machine.UART`? It's always worth being explicit about this (because the documentation and definition of the default pins are has changed over time).

FYI you should use `time.ticks_diff` for the timeout (See https://docs.micropython.org/en/latest/library/time.html#time.ticks_diff for reasons)

I'm not quite sure what this line is doing -- `resp = b"".join([resp, uart.read()])` -- did you mean to just write `resp += uart.read()` ?

The code otherwise looks like it should work so maybe there's something wrong with either the baud rate, the wiring, or the command? Is `AT` on its own a supported command? Maybe try one of the identification commands like `AT+GMI`.
